### PR TITLE
Break down kcidb.orm module

### LINKED ISSUE #386: Break down kcidb.orm module
The [`kcidb.orm`](https://github.com/kernelci/kcidb/blob/main/kcidb/orm/__init__.py) module is now 1623 lines long, which is a bit much. 

We can extract parts of it into separate modules to make it more readable and maintainable.

Extract the `SCHEMA` variable and its dependencies and assertions into a new module called `kcidb.orm.data`.

Extract the `Pattern` class and its dependencies into a new module called `kcidb.orm.query`.

Update all references as necessary.

### PR DESCRIPTION
**Pull request by octonawish-akcodes**
Extracted `Schema` and `Pattern` class and their dependencies and made them as a separate module from kcidb.orm module.
This PR closes #386 
@spbnick Need reviews.

### CODE REVIEW

File: kcidb/orm/data.py [hunk 1]
```diff
@@ -0,0 +1,302 @@
+from kcidb.orm.__init__ import Type, Relation,_DEFS, _CHECKOUT, _BUILD, _TEST, _INCIDENT, _ISSUE, _ISSUE_CULPRIT, _TEST_ENVIRONMENT
```
Comment (spbnick) [inline kcidb/orm/data.py:1]: This would create a circular dependency, and shouldn't happen.

We need to bring the definitions of these dependencies here as well (as I pointed out in the issue).

Additionally, please do not import from `__init__` directly. This creates a very tight "coupling" and makes the code more fragile.

File: kcidb/orm/query.py [hunk 2]
```diff
@@ -0,0 +1,1461 @@
+from kcidb.orm.__init__ import re, Type, LIGHT_ASSERTS, textwrap
```
Comment (spbnick) [inline kcidb/orm/query.py:1]: Same here: do not import anything from `__init__` explicitly. Also, definitely do not import common definitions available elsewhere (such as `LIGHT_ASSERTS`, or `textwrap`) via a third-party module.

**Review by spbnick — CHANGES_REQUESTED** [review]
Nice start, Abhishek! Left you a few comments/requests inline. Thanks!

**Comment by spbnick** [comment]
Don't hesitate to ask any questions you might have about this, here, or on Slack!

File: kcidb/orm/query.py [hunk 3]
```diff
@@ -0,0 +1,1461 @@
...
+from kcidb.orm.data import Schema, SCHEMA
+import kcidb.io as io
+import kcidb
```
[the remark this answers is not in the document]
Comment (octonawish-akcodes) [reply kcidb/orm/query.py:4]: got it.

File: kcidb/orm/query.py (hunk 2 above)
Comment (octonawish-akcodes) [reply kcidb/orm/query.py]: Okay sir.

File: kcidb/orm/data.py (hunk 1 above)
Comment (octonawish-akcodes) [reply kcidb/orm/data.py]: Okay, I will do that.

**Comment by octonawish-akcodes** [comment]
After making the suggested changes, Below is the output of the test cases. All are passing as they were passing before creating the separate modules for `Schema` and ` Pattern`.

image

**Comment by octonawish-akcodes** [comment]
In updated the PR and removed the reformatting, I'll make a separate PR for that if necessary. Now this PR is purely the break down changes to be reviewed @spbnick

**Review by spbnick — CHANGES_REQUESTED** [review]
Very nice! We're close to the finish line here!

Please deal with the CI errors, and from now on please make sure CI passes before review, as you're now more experienced with the project (please keep ignoring the deployment failures for now).

Please also don't forget to keep rebasing your changes on latest `main` as you work, and before reviews.

Finally, please explain a little what you're doing in the commit message body (take the issue description as a reference).

**Comment by octonawish-akcodes** [comment]
Test cases passed, fixes the cyclic imports, and updated the references where necessary. I also added a description of what I am doing and what I have done have a look, sir. This PR is intended to work as we wanted it to, need reviews @spbnick

File: kcidb/db/abstract.py [hunk 4]
```diff
@@ -1,7 +1,7 @@
 """Kernel CI reporting database - abstract database definitions"""
 
 from abc import ABC, abstractmethod
-import kcidb.orm as orm
+from kcidb.orm.query import Pattern
```
Comment (spbnick) [inline kcidb/db/abstract.py:4]: Could you instead write:
```python
import kcidb.orm.query
```
And refer to `Pattern` as `kcidb.orm.query.Pattern`? We only have one reference here (so there's not much typing/reading overhead) and it would be good to be clear where it's coming from exactly.

File: kcidb/orm/data.py [hunk 5]
```diff
@@ -0,0 +1,478 @@
+"""
+The kcidb.orm.data module defines the data structures and constants used
+by the KCIDB object-relational mapping system to organize Kernel CI report
+data into objects.
```
Comment (spbnick) [inline kcidb/orm/data.py:4]: Let's not put the module name into its description, so we don't forget to update it if we ever rename it. Besides, it's redundant. You know that it's a module, and which module it is when you're reading this. Otherwise this is a pretty good description, thanks! But let's start it with just "The data structures and constants ..." directly.

File: kcidb/orm/query.py [hunk 6]
```diff
@@ -0,0 +1,830 @@
+"""
+The kcidb.orm.query module defines the query classes and functions used
+by the KCIDB object-relational mapping system to retrieve Kernel CI report
+data from the database.
```
Comment (spbnick) [inline kcidb/orm/query.py:4]: Same here, let's just write "The query classes ..." and so on. Otherwise, another nice one!

**Review by spbnick — CHANGES_REQUESTED** [review]
Ah, very neat and tidy :relaxed: Thank you, Abhishek!

I only have a few minor requests left.

The commit message looks very good too, thanks!

It would be totally perfect if you used the imperative mood in the commit body as well (i.e. wrote "Extract parts ...", instead of "We extracted parts ...", and so on), but the most important part is the mood of the commit subject, and that's fine.

File: kcidb/db/abstract.py (hunk 4 above)
Comment (octonawish-akcodes) [reply kcidb/db/abstract.py]: Oh!!! Okay. Done.

File: kcidb/orm/query.py (hunk 6 above)
Comment (octonawish-akcodes) [reply kcidb/orm/query.py]: Done!!!

**Comment by octonawish-akcodes** [comment]
@spbnick Done all the requested changes, Need reviews.

File: kcidb/orm/__init__.py [hunk 7]
```diff
@@ -3,1287 +3,27 @@
...
-import textwrap
 import logging
-import argparse
 from abc import ABC, abstractmethod
-import jsonschema
 import kcidb.io as io
 import kcidb.misc
 from kcidb.misc import LIGHT_ASSERTS
+from kcidb.orm.query import Pattern, PatternHelpAction
+from kcidb.orm.data import SCHEMA
```
Comment (spbnick) [inline kcidb/orm/__init__.py:12]: Could you please import the `query` and `data` modules here, instead of the symbols directly? And refer to the symbols through them, e.g. `query.Pattern` and `data.SCHEMA`. First, this would satisfy #404, and then it would stop these symbols appearing like they're actually still defined here. It's best to always refer to the symbols by their canonical name and definition location, and not use third-party references (e.g. like the `Pattern` imported here would be).

In other words, the symbols you **moved** out of the `kcidb.orm` module should no longer be available there.

This will also require updating more code in other modules, including `main.py` in the top directory, containing the Cloud Function entry points.

File: kcidb/orm/__init__.py [hunk 8]
```diff
@@ -3,1287 +3,27 @@
...
-        ),
-    )
-)
-
 assert all(k.endswith("s") for k in io.SCHEMA.graph if k), \
     "Not all I/O object list names end with 's'"
 
 assert set(SCHEMA.types) >= \
     set(k[:-1] for k in io.SCHEMA.graph if k), \
     "OO types are not a superset of I/O types"
```
Comment (spbnick) [inline kcidb/orm/__init__.py:26]: These assertions should go into the `data` submodule, together with the `SCHEMA`, as they're closely tied to it, and are verifying either its pre-conditions, or the `SCHEMA` itself.

**Review by spbnick — CHANGES_REQUESTED** [review]
Thank you very much Abhishek! This looks very good and satisfies all my previous requests.

However, now that your PR is much neater I was able to notice one small, and one larger problem with this. I left you a couple requests inline.

Thank you.

File: kcidb/orm/__init__.py (hunk 7 above)
Comment (octonawish-akcodes) [reply kcidb/orm/__init__.py]: Okay

File: kcidb/orm/__init__.py (hunk 8 above)
Comment (octonawish-akcodes) [reply kcidb/orm/__init__.py]: done!!!

**Comment by spbnick** [comment]
@octonawish-akcodes, I tried reproducing your import problems, and it would go smoother if instead of things like this:
```python
import kcidb.orm.query as query
import kcidb.orm.query as data
```
you would write
```python
from kcidb.orm import query, data
```

**Comment by spbnick** [comment]
We don't have much time left for you to clean this up, and we're blocking #379, so I went ahead and did it for you. Please review the patch I added on top of yours, then squash them together, push to PR, and then review the result again.

Main changes I did:
* Remove all unnecessary changes, including whitespace and import changes. It's generally better to minimize churn and do only what's necessary for a particular task (provided it's done well). We can work on unifying and refining imports later.
* Update documentation so it refers to the new location of the moved symbols. Always remember to update the docs.
* Remove the `LOGGER` global from modules that don't need it. Do not add globals unnecessarily.

**Review by spbnick — APPROVED** [review]
LGTM. Thank you, @octonawish-akcodes. Merging!
